genetic_algorithms/algorithm_wrappers/algorithm_monitor.py

```diff
@@ -0,0 +1,53 @@
+from genetic_algorithms.algorithms.algorithm import AlgorithmConfig
+from re import S
+from genetic_algorithms.solvers.solver import SolverConfig
+from genetic_algorithms.problems.base.state import State
+from genetic_algorithms.problems.base.model import Model
+from genetic_algorithms.algorithm_wrappers.algorithm_wrapper import AlgorithmWrapper
+from rich.console import Console
+from rich.table import Table
+from rich.progress import Progress
+from time import sleep
+
+console = Console()
+
+
+class AlgorithmMonitor(AlgorithmWrapper):
+    def __init__(self, config: AlgorithmConfig, **kwargs):
+        super().__init__(**kwargs)
+        self._algorithm_config = config
+        self._best_algo_cost, self._iters_from_last_impr = float('inf'), 0
+        self._progress_bar = self._create_progress_bar()
+
+    def _create_step_table(self, new_cost):
+        table = Table(show_header=True, header_style="bold magenta")
+        columns = ["Title", "Value"]
+        for column in columns:
+            table.add_column(column)
+        table.add_row(
+            "Best found cost",
+            str(self._best_algo_cost)
+        )
+        table.add_row(
+            "Current cost",
+            str(new_cost)
+        )
+        return table
+
+    def _create_progress_bar(self):
+        completed = self._iters_from_last_impr
+        left = self._algorithm_config.max_steps_without_improvement - completed
+        return f'Steps from last impr: [[cyan]{"#" * completed}[/cyan]{"-" * (left)}]'
+
+    def _perform_side_effects(self, model: Model, state: State):
+        console.clear()
+        new_cost = model.cost_for(state)
+        if new_cost < self._best_algo_cost:
```

Comment:
To jeszcze trzeba będzie poprawić tak, żeby zależało od wybranej strategii optymalizacji.